---
title: "Test pseudotime"
subtitle: 'Fresh Microglia and NPSAD'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
    toc: true
    smart: false
---


<!--- 

# cd /Users/gabrielhoffman/workspace/repos/nps_ad/mg

cd /sc/arion/projects/CommonMind/hoffman/NPS-AD/work/nps_ad/mg
R --vanilla

system("ml git; git pull"); rmarkdown::render("test_pseudotime.Rmd");


# https://hoffmg01.hpc.mssm.edu/mg/test_pseudotime.html


--->



```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  cache = TRUE,
  cache.lazy = FALSE)
```


```{r load.packages}
library(tidyverse)
library(variancePartition)
library(lmerTest)
library(parallel)
library(ggplot2)
library(zellkonverter)
library(SingleCellExperiment)
```

```{r analysis}
# PsychAD MicroPVM + metadata
file = "/sc/arion/projects/Microglia/scRNAseq_proc/freeze3/5_taxonomy/1_scANVI_psychAD_annotation/230420_PsychAD_MicroPVM_independent_meta_scANVI_new-umap.h5ad"
# sce = readH5AD(file, use_hdf5=TRUE, verbose=TRUE, uns=FALSE, layers=FALSE, obsm=FALSE)
# saveRDS(colData(sce), file="/sc/arion/scratch/hoffmg01/data/colData_psychad.RDS")
sce_colData_psychad = readRDS("/sc/arion/scratch/hoffmg01/data/colData_psychad.RDS")

# Fresh MG
file = "/sc/arion/projects/Microglia/scRNAseq_proc/freeze3/4_qc/230424_FreshMG_freeze3_MicroPVM_anno_meta_HGNC_log1p.h5ad"
# sce2 = readH5AD(file, use_hdf5=TRUE, verbose=TRUE, uns=FALSE, layers=FALSE, obsm=FALSE, obsp=FALSE)
# saveRDS(colData(sce2), file="/sc/arion/scratch/hoffmg01/data/colData_fresh.RDS")
sce_colData_fresh = readRDS("/sc/arion/scratch/hoffmg01/data/colData_fresh.RDS")

files = dir("/sc/arion/scratch/hoffmg01/data/", pattern="*csv*",recursive=TRUE, full.names=TRUE)
names(files) = gsub(".csv", "", paste(basename(dirname(files)), basename(files), sep="/"))
 
dfList = lapply( names(files)[1:2], function(id){

	df = read_csv(files[id])
	# df$SubID =  gsub("-[GCTA]+", "", df$barcodekey)

	df2 = merge(df[,1:3], sce_colData_fresh, by.x="barcodekey", by.y="row.names")
	droplevels(as.data.frame(df2))
	})
names(dfList) = names(files)[1:2]

dfList2 = lapply( names(files)[3:4], function(id){

	df = read_csv(files[id])
	# colnames(df)[1] = "idx"

	# USe subtype 
	# df2 = cbind(df[,1:3], sce_colData_psychad[df$idx+1,])

	# droplevels(df2)
	df2 = merge(df[,1:3], sce_colData_psychad, by.x="barcodekey", by.y="row.names")
	droplevels(as.data.frame(df2))
	})
names(dfList2) = names(files)[3:4]

dfList = c(dfList, dfList2)
```

```{r plots, fig.width=12}
fitList = lapply(names(dfList), function(id){
	message(id)
	ggplot(dfList[[id]], aes(as.character(braak), dpt_pseudotime, fill=subtype)) +
	geom_violin() +
	theme_classic() +
	theme(aspect.ratio=1, legend.position="none") +
	xlab("Braak") + 
	ggtitle(id) +
	facet_wrap(~ subtype, nrow=2)
})
fitList
```


```{r lmer}
res = lapply(names(dfList), function(id){

	if( grepl("fresh_mg", id) ){
		form = dpt_pseudotime ~ braak + (1|subtype) + (1|Channel) + age + (1|sex) + PMI + log(n_genes) + (1|predicted_phase)
	}else{
		form = dpt_pseudotime ~ braak + (1|subtype) + (1|Channel) + age + (1|sex) + PMI + (1|poolID) + log(n_genes) + (1|predicted_phase)
	}

	data = dfList[[id]][,all.vars(form)]

	fit = lmer(form, data, REML=FALSE, control=lme4::lmerControl(optimizer="Nelder_Mead"))

	list(vp = calcVarPart(fit), summary = coef(summary(fit)))
})
names(res) = names(dfList)
```

```{r show.lmer, cache=FALSE}
res
```


# Analysis per cell type
```{r lmer.subtype}
traits = c("CERAD", "Braak", 'dementia', "AD_LessStrict_Feb23", "AD_Strict_Feb23")

res = lapply(names(dfList), function(id){

	data = dfList[[id]]

	res = lapply(as.character(unique(data$subtype)), function(st){
		res = lapply(traits, function(trait){

			if( grepl("fresh_mg", id) ){
				form = dpt_pseudotime ~ (1|Channel) + age + (1|sex) + PMI + log(n_genes) + (1|predicted_phase)
			}else{
				form = dpt_pseudotime ~ (1|Channel) + age + (1|sex) + PMI + (1|poolID) + log(n_genes) + (1|predicted_phase)
			}
			
			x = as.character(form)
			form = as.formula(paste(x[2], x[1], x[3], "+", trait))

			include = data$subtype == st

			if( !(trait %in% colnames(data)) ){
				return(NULL)
			}
			data2 = droplevels(data[include,all.vars(form)])

			fit = lmer(form, data2, control=lme4::lmerControl(optimizer="Nelder_Mead"))

			df = data.frame(coef(summary(fit)))
			df = data.frame(coef = rownames(df), df)

			data.frame( Dataset = id, 
						subtype = st,
						trait = trait,
						df[trait,])
		})
		do.call(rbind, res)
	})
	do.call(rbind, res)
})
res = do.call(rbind, res)
rownames(res) = c()
```



```{r plot.subtype, fig.height=12, fig.width=12}
res$pvalue = pmax(res[['Pr...t..']], 1e-10)

fitList = lapply(traits, function(trait){
	res[res$coef == trait,] %>%
		ggplot(aes(subtype, Estimate, fill=-log10(pvalue))) +
		geom_bar(stat="identity") +
		theme_classic() +
		theme(aspect.ratio=1) +
		facet_wrap(~ Dataset, nrow=2) +
		coord_flip() +
		scale_fill_gradient(name=bquote(-log[10]~p), low="black", high="red", limits=c(0, 10)) +
		ggtitle(trait)
})
fitList
```




Pseudotime

 test other phenotypes
 heatmap of mean PT over brack stage

 not interested in p-values
 	want to see which effect size is the largest

 restrict to Age > 40
 
 use robust regression or rank transform	




```{r exit, cache=FALSE, echo=FALSE}
knitr::knit_exit()
```


# Test pseudotime
# ```{r testPT}
# library(dreamlet)
# library(MASS)
# file = "/sc/arion/projects/Microglia/scRNAseq_proc/freeze3/5_taxonomy/1_scANVI_psychAD_annotation/230420_PsychAD_MicroPVM_independent_meta_scANVI_new-umap.h5ad"
# sce = readH5AD(file, use_hdf5=TRUE, verbose=TRUE, uns=FALSE, layers=FALSE, obsm=FALSE)

# counts(sce) = assay(sce, "X")
# logcounts(sce) = computeLogCPM(sce)

# colData(sce)$dpt_pseudotime = rep(NA, ncol(sce))

# idx = dfList[[3]]$idx + 1
# colData(sce)$dpt_pseudotime[idx] = dfList[[3]]$dpt_pseudotime

# idx = dfList[[4]]$idx + 1
# colData(sce)$dpt_pseudotime[idx] = dfList[[4]]$dpt_pseudotime

# sce_AH = sce[,sce$subclass %in% c('MG_Adapt', 'MG_Homeo')]
# sce_PVM = sce[,sce$subclass %in% c('MG_PVM')]
# ```

# ```{r make.plots, fig.height=12, fig.width=12}
# df = data.frame(colData(sce_AH), 
# 	gene = logcounts(sce_AH)["PTPRG",])

# df %>%
# 	filter( !is.na(dpt_pseudotime)) %>%
# 	ggplot(aes(dpt_pseudotime, gene, color=Braak)) +
# 	geom_point() +
# 	geom_smooth(method="lm", color="red") + 
# 	theme_classic() +
# 	theme(aspect.ratio=1) +
# 	ylab("PTPRG") +
# 	ggtitle("Adapt + Homeo") +
# 	facet_wrap(~ subtype)

# df = data.frame(colData(sce_PVM), 
# 	gene = logcounts(sce_PVM)["PTPRG",])

# df %>%
# 	filter(!is.na(dpt_pseudotime)) %>%
# 	ggplot(aes(dpt_pseudotime, gene, color=Braak)) +
# 	geom_point() +
# 	geom_smooth(method="lm", color="red") + 
# 	theme_classic() +
# 	theme(aspect.ratio=1) +
# 	ylab("PTPRG") +
# 	ggtitle("PVM") +
# 	facet_wrap(~ subtype)
# ```




# Fresh MG
# Test pseudotime
```{r testPT}
library(dreamlet)
library(MASS)
file = "/sc/arion/projects/Microglia/scRNAseq_proc/freeze3/4_qc/230424_FreshMG_freeze3_MicroPVM_anno_meta_HGNC_log1p.h5ad"
sce.in = readH5AD(file, use_hdf5=TRUE, verbose=TRUE, uns=FALSE, layers=FALSE, obsm=FALSE, obsp=FALSE, raw=TRUE)
sce = swapAltExp(sce.in, "raw")
rownames(sce) = rownames(sce.in)


counts(sce) = assay(sce, "X")
sce$lib.size = colSums2(counts(sce))

colData(sce)$dpt_pseudotime = rep(NA, ncol(sce))
colData(sce)$Dataset = rep(NA, ncol(sce))

idx = match(colnames(sce), dfList[[1]]$barcodekey)
colData(sce)$dpt_pseudotime[!is.na(idx)] = dfList[[1]]$dpt_pseudotime[idx[!is.na(idx)]]
colData(sce)$Dataset[!is.na(idx)]  = "Adapt_Homeo"

idx = match(colnames(sce), dfList[[2]]$barcodekey)
colData(sce)$dpt_pseudotime[!is.na(idx)] = dfList[[1]]$dpt_pseudotime[idx[!is.na(idx)]]
colData(sce)$Dataset[!is.na(idx)]  = "MG_PVM"

form = y ~ offset(log(lib.size)) + scale(dpt_pseudotime) + scale(age) + sex + scale(PMI) + scale(log(n_genes)) + (1|Channel) 

data = colData(sce)[,c('Dataset', 'subtype', all.vars(form)[-1])]
data$y = counts(sce)['PTPRG',]


data2 = data[which(data$Dataset == 'Adapt_Homeo'),]
data2 = data2[data2$subtype == 'MG_Adapt_AIF1',]
data2 = as.data.frame(data2)

fit = glmer.nb(form, data2)

library(glmmTMB)
form = y ~ offset(log(lib.size)) + scale(dpt_pseudotime) + scale(age) + sex + scale(PMI) + scale(log(n_genes)) + (1|Channel) 

fit2 = glmmTMB(form, data2, family=nbinom1)

library(NBZIMM)
# f1 = glmm.nb(y ~ Days + Age + Race + preg + offset(log(N)), random = ~ 1 | subject)

form = y ~ scale(dpt_pseudotime) + scale(age) + sex + scale(PMI) + scale(log(n_genes)) + offset(log(lib.size))
fit3 = glmm.nb(form, random = ~ 1|Channel, data)



data = colData(sce)
data$y = counts(sce)['PTPRG',]
form = y ~ offset(log(lib.size)) + PMI
df = as.data.frame(data[,all.vars(form)])


fit = glm.nb(form, df)

summary(fit)






```









<!---

PsychAD metadata
/sc/arion/projects/psychAD/NPS-AD/freeze2_rc/metadata/clinical_metadata.csv
/sc/arion/projects/psychAD/NPS-AD/freeze2_rc/metadata/clinical_metadata_sampleSets_latest.RDS
--->

